### PR TITLE
replace clear_mutable_collection with a data structure

### PR DESCRIPTION
**Pull request by nikitaved**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #607

### CODE REVIEW

File: thunder/__init__.py [hunk 1]
```diff
@@ -621,8 +621,6 @@ def get_computation_and_inputs(*args, **kwargs):
             # See https://github.com/Lightning-AI/lightning-thunder/issues/433
             if cd.use_cudagraphs:
                 comp = CUDAGraphExecutor(comp)
-                if backward_fn is not None:
-                    backward_fn = CUDAGraphExecutor(backward_fn, num_constant_args=len(backward_trc.args[0][0]))
```
Comment (IvanYashchuk) [inline thunder/__init__.py:625]: Will you restore this code?

Comment (nikitaved) [reply thunder/__init__.py:625]: Yes, in a follow-up stack PRs.

File: thunder/executors/torch_autograd.py [hunk 2]
```diff
@@ -60,6 +60,41 @@ def rename_bwd_trace_outputs(bwd_trace: TraceCtx, fwd_trace: TraceCtx) -> TraceC
...
+
+        while isinstance(t, BasicTensorSubclass):
+            t = t.tensor_obj
+
+        res = torch.Tensor._make_subclass(cls, t, t.requires_grad)
+        res.tensor_obj = t
+        return res
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
```
[the remark this answers is not in the document]
Comment (nikitaved) [reply thunder/executors/torch_autograd.py:84]: As long as it has `data`, it is fine for nvfuser.

File: thunder/__init__.py (hunk 1 above)
Comment (IvanYashchuk) [reply thunder/__init__.py:625]: Is there a way to not disable functionality in one piece of the stack? What tests fail?

File: thunder/executors/torch_autograd.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (IvanYashchuk) [reply thunder/executors/torch_autograd.py:84]: Is it the case for all functions that expect torch.Tensor as input? Why is `__torch_dispatch__` needed then?

File: thunder/__init__.py (hunk 1 above)
Comment (nikitaved) [reply thunder/__init__.py:625]: It is not meant to touch this part just now, doing so for testing purposes. The correct answer to your top question is "it needs modifications". This PR, however, will not touch it just yet.

File: thunder/executors/torch_autograd.py [hunk 3]
```diff
@@ -60,6 +60,41 @@ def rename_bwd_trace_outputs(bwd_trace: TraceCtx, fwd_trace: TraceCtx) -> TraceC
...
+    """
+    A tensor subclass that does own explicit storage (created with _make_wrapper_subclass),
+    has a reference to a torch.Tensor instance, and dispatches to torch.Tensor operations.
+    """
+
+    @staticmethod
+    def __new__(cls, t: torch.Tensor | None):
+        # Allow to propagate None - this is useful in the context of torch.autograd
+        if t is None:
+            return None
```
Comment (IvanYashchuk) [inline thunder/executors/torch_autograd.py:74]: saved_tensors is implemented to store only tensors and nothing else is allowed by torch.autograd.Function's ctx.save_for_backward method.

Comment (nikitaved) [reply thunder/executors/torch_autograd.py:74]: Some failing tests begged to differ

File: thunder/executors/torch_autograd.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nikitaved) [reply thunder/executors/torch_autograd.py:84]: `__torch_dispatch__` is needed for PyTorch to understand subclasses. It is being called only once per each wrapped tensor.

**Comment by nikitaved** [comment]
Closing - https://github.com/Lightning-AI/lightning-thunder/pull/656 solves it in a more Thunder-appropriate way.

**Pull request closed without merging**